misk/src/main/kotlin/misk/digester/MergingDigest.kt

```diff
@@ -0,0 +1,323 @@
+package misk.digester
+
+import tdigest.Centroid
+import tdigest.MergingDigestData
+import kotlin.math.PI
+import kotlin.math.asin
+
+/**
+ * Provides an implementation of t-digest to be used to send histogram data
+ *
+ * Direct port of Veneur digest created by Stripe
+ * https://github.com/stripe/veneur/blob/master/tdigest/merging_digest.go
+ */
+class MergingDigest {
+
+  // amount of compression to data
+  private val compression: Double
+
+  // main list of centroids
+  private var mainCentroids: MutableList<Centroid> = mutableListOf()
+  // total weight of unmerged main centroids
+  private var mainWeight: Double = 0.0
+
+  // centroids that have been added but not yet merged into main list
+  private var tempCentroids: MutableList<Centroid> = mutableListOf()
+  // total weight of unmerged temp centroids
+  private var tempWeight: Double = 0.0
+
+  private var min: Double = Double.POSITIVE_INFINITY
+  private var max: Double = Double.NEGATIVE_INFINITY
+
+  constructor(compression: Double) {
+    this.compression = compression
+  }
+
+  /**
+   * Constructs a MergingDigest with values initialized from MergingDigestData.
+   * This should be the way to generate a MergingDigest from a serialized protobuf.
+   */
+  constructor(mergingDigestData: MergingDigestData) {
+    compression = mergingDigestData.compression
+    mainCentroids = mergingDigestData.main_centroids.toMutableList()
+    min = mergingDigestData.min
+    max = mergingDigestData.max
+
+    mainWeight = 0.0
+    for (c in mainCentroids) {
+      mainWeight += c.weight
+    }
+  }
+
+  /**
+   * Returns the estimated size of the temporary buffer based on compression value
+   *
+   * Function taken from stripes implementation of merging_digest
+   * https://github.com/stripe/veneur/blob/master/tdigest/merging_digest.go
+   */
+  private fun estimateTempBuffer(compressionValue: Double): Int {
+    // this heuristic comes from Dunning's paper
+    // 925 is the maximum point of this quadratic equation
+    val tempCompression = minOf(925.0, maxOf(20.0, compressionValue))
+    return (7.5 + 0.37*tempCompression - 2e-4*tempCompression*tempCompression).toInt()
+  }
+
+  /**
+   * Adds a new value to the t-digest, with a given weight that must be positive.
+   * Infinities and NaN cannot be added.
+   */
+  fun add(value: Double, weight: Double) {
+    require (value != Double.NaN && value != Double.POSITIVE_INFINITY && weight > 0) {
+      "invalid value added"
+    }
+
+    if (tempCentroids.size == estimateTempBuffer(compression)) {
```

Comment:
Formatting is off on all of these files (should be two space indent). Make sure you to use the Square Intellij code style settings (https://github.com/square/java-code-styles)